[16.0][IMP] sale_order_lot_selection: add lot selection wizard

halbtonjazz opened the pull request:
Implement lot selection wizard for sales quotations to enable creating quotations with specific lot/serial numbers directly from stock lots. This improves traceability and allows precise lot selection for industries requiring specific serial number tracking.

Task: 4470

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced an interactive wizard to add stock lots to sale orders, supporting both new creations and updates.
  - Added a dedicated menu item for easy access to lot and serial number management.
  - Enabled a configuration option to generate sales quotations directly from stock lots.

- **Documentation**
  - Enhanced guides with expanded sections outlining real-world use cases, detailed workflows, and step-by-step instructions for managing lot and serial number specifications.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

halbtonjazz committed:
[IMP] sale_order_lot_selection: add lot selection wizard

Implements a wizard interface for managing lot selection in sale orders:
- Add configuration option to enable/disable lot selection from stock
- Introduce wizard to select multiple lots with quantities
- Add security rules and access rights for lot operations
- Enhance UX with dynamic quantity validation against available stock

Task: 4470

halbtonjazz committed:
[IMP] sale_order_lot_selection: add lot selection wizard

- Refactor the wizard to handle new instances, improving reliability.
- Add validations to ensure a sale order or customer is selected, preventing errors.
- Update the UI to reflect new functionality and provide better user guidance.
- Set default lot quantities based on product quantities for accuracy.

Task: 4470

halbtonjazz committed:
[ADD] sale_order_lot_selection: add tests

Add test cases to verify functionality of creating sale orders from stock
lots and adding lots to existing orders

Task: 4470

halbtonjazz committed:
[IMP] sale_order_lot_selection: update docs

Add copyright headers to Python and XML files.
Update usage instructions to reflect the new lot selection wizard
feature and configuration options.
Add Cetmix contributors and new maintainer to module metadata.

Task: 4470

halbtonjazz committed:
[ADD] sale_order_lot_selection: add use cases

Add business use cases and context to the documentation to explain
why selecting specific lot or serial numbers when creating
quotations is needed.
Describes scenarios where product traceability, compliance, or
customer-specific requirements are essential.

Task: 4470

halbtonjazz committed:
[REF] sale_order_lot_selection: refactor wizard

Remove maintainer key as modules are not actively maintained.
Update view name in res_config_settings to better reflect module purpose.
Apply code review suggestions from coderabbitai bot.

Task: 4470

CetmixGitDrone commented:
Hi @bodedra,
some modules you are maintaining are being modified, check this out!

halbtonjazz commented on sale_order_lot_selection/readme/CONTEXT.rst:
```diff
@@ -0,0 +1,20 @@
+specifications before confirming orders, such as used equipment with unique serial
+numbers or specific inventory lots.
+
+This is particularly important in several business scenarios:
+
+* In industries where product traceability is critical and customers need specific
+  serial numbers (electronics, automotive parts, medical devices)
+* When dealing with batches that vary in quality or characteristics, where
+  customers require products from specific production runs
+* In regulated industries where lot traceability is mandatory for compliance
+  purposes
+
```
Current format follows OCA standards and is sufficiently readable

halbtonjazz commented on sale_order_lot_selection/readme/CONTEXT.rst:
```diff
@@ -0,0 +1,20 @@
+* In regulated industries where lot traceability is mandatory for compliance
+  purposes
+
+For example:
+
+* A machinery reseller needs to quote specific used equipment with known serial
+  numbers
+* An electronics distributor must sell particular serial-numbered items for
+  warranty tracking
+* A car parts supplier needs to specify exact serial numbers for rare components
```
Same as above - current format follows OCA standards and is clear enough

halbtonjazz commented on sale_order_lot_selection/tests/__init__.py:
```diff
@@ -1,4 +1,6 @@
 # © 2015 Agile Business Group
+# Copyright (C) 2025 Cetmix OÜ
```
The copyright year 2025 is correct as it reflects the current year

halbtonjazz commented on sale_order_lot_selection/models/res_config_settings.py:
```diff
@@ -0,0 +1,14 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
The copyright year 2025 is correct as it reflects the current year

halbtonjazz commented on sale_order_lot_selection/__manifest__.py:
```diff
@@ -1,3 +1,6 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
The copyright year 2025 is correct as it reflects the current year

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        for record in self:
+            record.partner_id = record.sale_order_id.partner_id.id
+
+    def open_wizard(self, lot_ids):
+        """Open the wizard with selected lots."""
+        if (
+            not self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("sale_order_lot_selection.allow_generate_from_lots")
+        ):
+            raise AccessError(_("You are not allowed to generate Sale Order from Lot."))
+        vals = {
+            "line_ids": [
+                (
+                    0,
+                    0,
+                    {
+                        "lot_id": lot_id.id,
+                        "quantity": lot_id.product_qty,
+                    },
+                )
+                for lot_id in lot_ids
+            ]
```
Current docstring follows OCA style guidelines. The parameter name is self-explanatory and the method's purpose is clearly stated

halbtonjazz commented on sale_order_lot_selection/static/description/index.html:
```diff
@@ -375,68 +375,123 @@
 The removal of this feature was made on this commit : 3234b2ccccf1dffafbe0a8fa8efaaea44f2e47ef.</p>
 </div>
 <div class="section" id="bug-tracker">
-<h1><a class="toc-backref" href="#toc-entry-3">Bug Tracker</a></h1>
+<h1><a class="toc-backref" href="#toc-entry-5">Bug Tracker</a></h1>
 <p>Bugs are tracked on <a class="reference external" href="https://github.com/OCA/sale-workflow/issues">GitHub Issues</a>.
 In case of trouble, please check there if your issue has already been reported.
 If you spotted it first, help us to smash it by providing a detailed and welcomed
 <a class="reference external" href="https://github.com/OCA/sale-workflow/issues/new?body=module:%20sale_order_lot_selection%0Aversion:%2016.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**">feedback</a>.</p>
 <p>Do not contact contributors directly about support or help with technical issues.</p>
 </div>
```
This is an auto-generated file from RST documentation using standard OCA template. We should not modify it directly

halbtonjazz commented on sale_order_lot_selection/view/res_config_settings.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="res_config_settings_view_form_stock" model="ir.ui.view">
+        <field
+            name="name"
+        >res.config.settings.view.form.inherit.sale_order_lot_selection</field>
+        <field name="model">res.config.settings</field>
+        <field name="inherit_id" ref="stock.res_config_settings_view_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//div[@id='warning_info']" position="after">
+                <div
+                    class="col-12 col-lg-6 o_setting_box"
+                    groups="stock.group_production_lot"
+                >
+                    <div class="o_setting_left_pane">
+                        <field name="is_allow_generate_from_lots" />
+                    </div>
+                    <div class="o_setting_right_pane">
+                        <label for="is_allow_generate_from_lots" />
+                        <div class="text-muted">
+                            Allow to generate Sale Quotations from stock lots
+                        </div>
+                    </div>
+                </div>
+            </xpath>
+        </field>
+    </record>
+</odoo>
```
Thanks for the naming convention suggestion. I'll discuss this with the reviewers team to align with their preferences

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.xml:
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!--
+    Copyright (C) 2025 Cetmix OÜ
+    License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+-->
```
The copyright year 2025 is correct as it reflects the current year

halbtonjazz commented on sale_order_lot_selection/tests/test_stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,132 @@
+        wizard.action_add_lots_to_sale_order()
+
+        sale_order = self.sale_empty
+        lines = sale_order.order_line.filtered(lambda l: l.lot_id == self.lot_cable)
```
Thanks for the suggestion. While more descriptive names can improve readability, in this case the single-letter lambda parameter is a common pattern in Odoo codebase, especially for short filter expressions

halbtonjazz commented on sale_order_lot_selection/static/description/index.html:
```diff
@@ -375,68 +375,123 @@ <h1 class="title">Sale Order Lot Selection</h1>
+<li>Under Operations section select the “Allow to generate sales quotations from stock lots” check box</li>
+</ul>
+</div>
+<div class="section" id="usage">
+<h1><a class="toc-backref" href="#toc-entry-3">Usage</a></h1>
+<p>There are two ways to add Lots to a Sale Order:</p>
+<ol class="arabic simple">
+<li>From Sale Order:<ul>
+<li>Create/edit a product and set traceability by ‘By Lots’ option</li>
+<li>Create a new lot number and assign product</li>
+<li>Update quantity for that product and assign lot number</li>
+<li>Go to Sales &gt; Orders &gt; Quotations</li>
+<li>Create a new quotation and add recently above configured product</li>
+<li>Select lot number and confirm it</li>
 <li>Delivery order will reserve the lot when available</li>
 </ul>
+</li>
+<li>From Lots/Serial Numbers:<ul>
+<li>Go to “Sales &gt; Products &gt; Lots/Serial Numbers”</li>
+<li>Select Serial Numbers you would like to add to quotation</li>
+<li>Click action &gt; Create quotation</li>
+<li>In the action wizard:<ul>
+<li>To add to existing order: Select a Quotation/Sales order and click “Add”</li>
+<li>To create new order: Click “New” to create quotation with selected lots</li>
+</ul>
+</li>
+</ul>
+</li>
+</ol>
 </div>
```
This is an auto-generated file from RST documentation using standard OCA template. We should not modify it directly

halbtonjazz commented on sale_order_lot_selection/README.rst:
```diff
@@ -36,16 +36,62 @@ This selected lot number will be the one delivered to the Customer.
+Configuration
+=============
+
+If you want to allow users to generate sales quotas directly from the stock.lots tree view:
```
Thanks for catching this terminology difference. @Aldeigja could you please confirm which term we should use here - "quotas" or "quotations"?

halbtonjazz commented on sale_order_lot_selection/static/description/index.html:
```diff
@@ -375,68 +375,123 @@
+</ol>
 </div>
 <div class="section" id="known-issues-roadmap">
-<h1><a class="toc-backref" href="#toc-entry-2">Known issues / Roadmap</a></h1>
+<h1><a class="toc-backref" href="#toc-entry-4">Known issues / Roadmap</a></h1>
 <p>Block sale order validation on confirmation if the selected lot has been removed of this module
 Indeed nobody seems to know  what it had been implemented on early versions and it is really blocking other generic use case.
 One may want to validate a sale order restricting it to a lot that will be produced for instance and does not exist yet.</p>
 <p>This feature may be easyly put it back, in a separated module, if a use case requires it.
 The removal of this feature was made on this commit : 3234b2ccccf1dffafbe0a8fa8efaaea44f2e47ef.</p>
 </div>
```
This is an auto-generated file from RST documentation using standard OCA template. We should not modify it directly

GabbasovDinar commented:
hello @halbtonjazz this repo is fork from OCA, you need create PR to OCA repo from this fork

GabbasovDinar commented on sale_order_lot_selection/models/res_config_settings.py:
```diff
@@ -0,0 +1,14 @@
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    is_allow_generate_from_lots = fields.Boolean(
```
just `allow_generate_so_from_lots`

halbtonjazz replied:
Fixed as suggested. Thanks

GabbasovDinar commented on sale_order_lot_selection/models/sale_order_line.py:
```diff
@@ -1,3 +1,6 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
not need

halbtonjazz replied:
Understood, I added the header to comply with OCA guidelines,
but since I didn’t modify this file, I’ll remove it.

GabbasovDinar commented on sale_order_lot_selection/tests/test_stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,132 @@
+
+
+class TestAddLotToSaleOrder(TransactionCase):
+    def setUp(self):
+        super().setUp()
+        self.sale = self.env.ref("sale_order_lot_selection.sale1")
+        self.sale_empty = self.env["sale.order"].create(
+            {
+                "partner_id": self.sale.partner_id.id,
+                "user_id": self.env.ref("base.user_admin").id,
+            }
+        )
+        self.lot_cable = self.env.ref("sale_order_lot_selection.lot_cable")
+        self.env["ir.config_parameter"].sudo().set_param(
+            "sale_order_lot_selection.allow_generate_from_lots", True
+        )
```
from this [doc](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0): If there's a test class using `setUp` to generate test records, move it to `setUpClass` as it's way more performant since it runs only once.

halbtonjazz replied:
Fixed. I initially followed the pattern from test_sale_order_lot_selection.py, but moved setup to setUpClass for better performance as suggested

GabbasovDinar commented on sale_order_lot_selection/tests/test_stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,132 @@
+
+    def test_create_sale_order_with_lots(self):
+        """Test creating a new sale order from selected lot via wizard."""
+        wizard = self.env["stock.lot.add.to.sale.order.wizard"].create(
+            {
+                "line_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "lot_id": self.lot_cable.id,
+                            "quantity": self.lot_cable.product_qty,
+                        },
+                    )
+                ]
+            }
+        )
+        wizard.partner_id = self.sale.partner_id
```
```suggestion
        wizard = self.env["stock.lot.add.to.sale.order.wizard"].create(
            {
                "partner_id": self.sale.partner_id.id,
                "line_ids": [
                    (
                        0,
                        0,
                        {
                            "lot_id": self.lot_cable.id,
                            "quantity": self.lot_cable.product_qty,
                        },
                    )
                ]
            }
        )
```

halbtonjazz replied:
Thanks, I tried passing partner_id directly in create,
but reverted the change as it breaks test_error_if_no_partner_on_create which specifically tests validation when partner is missing. The separate partner_id assignment is needed for this test case.

GabbasovDinar commented on sale_order_lot_selection/tests/test_stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,132 @@
+        self.assertEqual(sale_order.order_line.lot_id, self.lot_cable)
+        self.assertEqual(sale_order.order_line.product_id, self.lot_cable.product_id)
+
+    def test_add_lots_to_existing_sale_order(self):
```
please, add docstring

halbtonjazz replied:
done

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+from odoo.exceptions import AccessError, ValidationError
+
+
+class StockLotAddToSaleOrder(models.TransientModel):
+    _name = "stock.lot.add.to.sale.order.wizard"
```
the model name and class name should me equal: 

```python
class StockLotSaleOrderWizard(models.TransientModel):
    _name = "stock.lot.sale.order.wizard"
```

halbtonjazz replied:
Thanks, I fixed it

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        }
+
+
+class StockLotAddToSaleOrderLine(models.TransientModel):
```
please move the models to separate file: `stock_lot_sale_order_line_wizard.py`


```python
class StockLotSaleOrderLineWizard(models.TransientModel):
    _name = "stock.lot.sale.order.line.wizard"
```

halbtonjazz replied:
done

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+                )
+
+
+class StockLot(models.Model):
+    _inherit = "stock.lot"
```
to separate file, it's not a wizard:

```
-> models/stock_lot.py
```

halbtonjazz replied:
Thank you for the review. I have moved the models to separate files as suggested

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        for record in self:
+            record.partner_id = record.sale_order_id.partner_id.id
+
+    def open_wizard(self, lot_ids):
```
this method should not be a method of the wizard class, move this method to the `StockLot` class

GabbasovDinar replied:
rename to: 

```python

def action_generate_sale_order(self):

```

halbtonjazz replied:
Done

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        self.sale_order_id = self.env["sale.order"].create(
+            {
+                "partner_id": self.partner_id.id,
+                "order_line": self._get_vals_for_add_lot_in_sale_order_line(),
```
`self._get_vals_for_add_lot_in_sale_order_line -> self._prepare_sale_order_line_vals`

halbtonjazz replied:
Done

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+                    "product_uom_qty": line_id.quantity,
+                },
+            )
+            for line_id in self.line_ids
```
```suggestion
            for line in self.line_ids
```

halbtonjazz replied:
Ok, I get why it's better to replace with line

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+            for line_id in self.line_ids
+        ]
+
+    def _action_open_sale_order(self):
```
why is private method?

halbtonjazz replied:
show other developers that this method is designed to be called inside this model only

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+    def _action_open_sale_order(self):
+        """Open the sale order form view."""
+        return {
+            "type": "ir.actions.act_window",
+            "name": "Sale Order",
+            "res_model": "sale.order",
+            "view_mode": "form",
+            "res_id": self.sale_order_id.id,
+            "target": "current",
```
use exist sale order action: ```self.env.ref("sale.<EXIST ACTION XML ID>")```

halbtonjazz replied:
I haven't found a suitable example of action in the odoo sale module. Could you be more specific about which action is better to use in this case, please? And why is it bad to create an action in python code through a dictionary for a specific task?

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+        store=True,
+        compute="_compute_partner_id",
+    )
+
```
add new field: 

```python

lot_ids = fields.Many2many("stock.lot", required=True)

```

halbtonjazz replied:
We store lots in the line model. Why should we store them in two places?

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+    partner_id = fields.Many2one(
+        "res.partner",
+        string="Customer",
+        store=True,
+        compute="_compute_partner_id",
```
no need, use `inverse` method for sale_order_id, e.g.:

```python
sale_order_id = fields.Many2one(
        "sale.order",
        domain="[('state', 'not in', ['done', 'cancel'])]",
        inverse="_inverse_sale_order_id",
)

def _inverse_sale_order_id(self):
    for record in self:
        if not record.sale_order_id:
            continue
        record.partner_id = record.sale_order_id.partner_id.id
```

halbtonjazz replied:
Why can't we use compute here? 
Does inverse recalculate the value dynamically on the interface?

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+class StockLot(models.Model):
+    _inherit = "stock.lot"
+
+    def action_add_to_sale_order(self):
+        """Open the wizard to add lots to a sale order."""
+        self.ensure_one()
+        return self.env["stock.lot.add.to.sale.order.wizard"].open_wizard(self)
```
```python
def action_generate_sale_order(self):
    if (
            not self.env["ir.config_parameter"]
            .sudo()
            .get_param("sale_order_lot_selection.allow_generate_from_lots")
        ):
            raise ValidationError(_("You are not allowed to generate Sale Order from Lot. Please check settings to allow generate sale order"))

        context = self.env.context.copy()
        context["lot_ids"] = self.ids
        return {
            "type": "ir.actions.act_window",
            "name": _("Generate Sale Order"),
            "res_model": "stock.lot.sale.order.wizard",
            "view_mode": "form",
            "context": context,
            "target": "new",
        }

```

halbtonjazz replied:
I understand the pros of this approach. It doesn't create a record in the database in advance. But isn't it a complication?
Since the wizard already creates lines with lot_id and quantity, adding a lot_ids field would duplicate that data and require additional logic to sync both sides. Creating the wizard directly with .create(...) is a common and robust practice in Odoo for such use cases with One2many lines. So, I suggest keeping the current implementation using `create()` because:
1. It follows standard Odoo patterns for wizard creation
2. Maintains a single source of truth (only `line_ids`)
3. Provides clearer data flow and easier debugging
4. Less potential errors and avoids potential synchronization issues between `lot_ids` and `line_ids`

The proposed context-based approach would require additional complexity and implementation time. Let's keep it simple and maintainable.
Let me know if you still see any particular issue with the current approach

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,161 @@
+    def action_add_lots_to_sale_order(self):
+        """Add selected lots to the existing sale order."""
+        self.ensure_one()
+        if not self.sale_order_id:
+            raise ValidationError(_("Please select a Sale Order"))
```
why? 

`Select an existing sale order or leave empty to create a new one`

halbtonjazz replied:
For the “add” action, the user has to select SO. For the “create” action, the user does not have to select an SO

GabbasovDinar commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.xml:
```diff
@@ -0,0 +1,61 @@
+        <field name="binding_view_types">tree</field>
+        <field name="state">code</field>
+        <field name="code">
+    if records:
+        action = env["stock.lot.add.to.sale.order.wizard"].open_wizard(records)
```
```suggestion
    if records:
        action = records.action_generate_sale_order()
```

halbtonjazz replied:
I fixed that part

ivs-cetmix commented:
@halbtonjazz test are falling because there are "rebel" [modules](https://github.com/cetmix/oca-sale-workflow/blob/d3b48a1cb3d67624e0b23a06f76316014131e6ee/.github/workflows/test.yml#L46) in this repo.
Those are modules that are not compatible with other modules in the same repo, so the same tests are run twice: for all modules and for the "rebel" ones. This is why a pair of test always shown as failed. This is not an issue.

halbtonjazz committed:
[REF] sale_order_lot_selection: improve SO generation flow

Restructure wizard implementation for better maintainability:
- Simplify model names and method signatures for clarity
- Split wizard line models into separate files
- Standardize configuration parameter names
- Update tests to cover refactored functionality
- Enhance UX for generating sale orders from stock lots

The changes make the code more organized and easier to maintain while
improving the user experience when creating sales from lot selection.

Task: 4470

halbtonjazz commented:
> hello @halbtonjazz this repo is fork from OCA, you need create PR to OCA repo from this fork
@GabbasovDinar ok, I got it. I will open a new pr after closing all unresolved conversations

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_sale_order_wizard_line.py:
```diff
@@ -0,0 +1,38 @@
+        string="Lot",
+        required=True,
+    )
+    product_id = fields.Many2one(related="lot_id.product_id")
+    quantity = fields.Float(string="Quantity")  # pylint: disable=W8113
+    max_qty = fields.Float(string="Max Quantity", related="lot_id.product_qty")
```
Thanks for the suggestion, but setting a default quantity value could be misleading as users should explicitly specify how many units they want to order from each lot to avoid unintended quantities in sale orders

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_sale_order_wizard_line.py:
```diff
@@ -0,0 +1,38 @@
+    quantity = fields.Float(string="Quantity")  # pylint: disable=W8113
+    max_qty = fields.Float(string="Max Quantity", related="lot_id.product_qty")
+
+    @api.constrains("quantity")
+    def _check_quantity(self):
+        """Check that the quantity is not greater than the max quantity."""
+        for record in self:
+            if record.quantity > record.max_qty:
+                raise ValidationError(
+                    _(
+                        "The quantity of lot %(lot_name)s cannot be greater than "
+                        "the available quantity %(max_qty)s.",
+                        lot_name=record.lot_id.name,
+                        max_qty=record.max_qty,
+                    )
+                )
```
Let me check how quantity validation currently works in the UI and discuss this with the reviewer, as there might be existing field-level validation handling this case.

halbtonjazz commented on sale_order_lot_selection/models/stock_lot.py:
```diff
@@ -0,0 +1,48 @@
+class StockLot(models.Model):
+    _inherit = "stock.lot"
+
+    def action_add_to_sale_order(self):
+        """Open the wizard to add lots to a sale order."""
+        self.ensure_one()
+        return self.action_generate_sale_order()
```
This method is specifically for the context menu action on a single lot record, hence the ensure_one() check is correct here.

halbtonjazz commented on sale_order_lot_selection/models/stock_lot.py:
```diff
@@ -0,0 +1,48 @@
+        self.ensure_one()
+        return self.action_generate_sale_order()
+
+    def action_generate_sale_order(self):
+        """Open wizard to generate sale order from lots."""
+        if (
+            not self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("sale_order_lot_selection.allow_generate_from_lots")
+        ):
+            raise AccessError(_("You are not allowed to generate Sale Order from Lot."))
+
```
AccessError is correct here as it's a permission restriction, not a data validation issue.

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.xml:
```diff
@@ -0,0 +1,61 @@
+        <field name="name">stock.lot.sale.order.wizard.form</field>
+        <field name="model">stock.lot.sale.order.wizard</field>
+        <field name="arch" type="xml">
+            <form string="Add Lot to Sale Order" create="false">
```
The create="false" is correct here as this wizard should only be launched from the lot action, never created manually.

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.xml:
```diff
@@ -0,0 +1,61 @@
+            <form string="Add Lot to Sale Order" create="false">
+                <group>
+                    <field
+                        name="sale_order_id"
+                        placeholder="Select an existing sale order or leave empty to create a new one"
+                    />
```
The current placeholder text is clear and descriptive, no need to sacrifice clarity for brevity here.

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,92 @@
+        )
+        return self._action_open_sale_order()
+
+    def _prepare_sale_order_line_vals(self):
+        """Get values for adding lot in sale order line."""
+        if not self.line_ids:
+            raise ValidationError(_("Please select at least one lot."))
+        return [
+            (
+                0,
+                0,
+                {
+                    "lot_id": line.lot_id.id,
+                    "product_id": line.lot_id.product_id.id,
+                    "product_uom_qty": line.quantity,
+                },
+            )
+            for line in self.line_ids
+        ]
+
```
This validation is unnecessary since users might want to add multiple lines with the same lot (e.g., for different prices or delivery terms), as shown in test_01_several_lines_with_same_lot.

halbtonjazz commented on sale_order_lot_selection/wizards/stock_lot_add_to_sale_order.py:
```diff
@@ -0,0 +1,92 @@
+        domain="[('state', 'not in', ['done', 'cancel'])]",
+    )
+
+    partner_id = fields.Many2one(
+        "res.partner",
+        string="Customer",
+        store=True,
+        compute="_compute_partner_id",
+    )
+
+    @api.depends("sale_order_id")
+    def _compute_partner_id(self):
+        for record in self:
+            record.partner_id = record.sale_order_id.partner_id.id
```
No inverse needed - partner_id is computed from sale_order_id, so it should only change when sale_order_id changes.
